File: app/test/features/sync_test.dart
import 'package:dartchess/dartchess.dart' show Side;
import 'package:drift/drift.dart' show Value;
import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:masteropening/core/db/app_database.dart';
import 'package:masteropening/core/db/daos/key_value_dao.dart';
import 'package:masteropening/core/db/enums.dart';
import 'package:masteropening/features/sync/data/session_store.dart';
import 'package:masteropening/features/sync/data/sync_api.dart';
import 'package:masteropening/features/sync/data/sync_repository.dart';
import 'package:masteropening/features/sync/domain/sync_payload.dart';

/// Ein Server im Arbeitsspeicher, der sich an dieselbe Konfliktregel hält.
class _FakeApi extends SyncApi {
  _FakeApi() : super(baseUrl: 'https://test.invalid');

  final Map<String, RepertoireDto> repertoires = {};
  final Map<String, ProgressDto> progress = {};
  ProfileDto? profile;

  DateTime serverTime = DateTime.utc(2026, 7, 30, 12);

  PushPayload? lastPush;
  DateTime? lastSince;
  int refreshCalls = 0;

  /// Lehnt das erste Zugriffstoken ab — dann muss der Aufrufer erneuern.
  bool rejectFirstCall = false;
  bool _rejected = false;

  void _checkToken() {
    if (rejectFirstCall && !_rejected) {
      _rejected = true;
      throw const SyncException('abgelaufen', isAuthFailure: true);
    }
  }

  @override
  Future<SyncSession> signInWithLichess(String lichessToken) async =>
      const SyncSession(accessToken: 'zugriff', refreshToken: 'erneuerung');

  @override
  Future<SyncSession> refresh(String refreshToken) async {
    refreshCalls++;
    return const SyncSession(
      accessToken: 'zugriff-neu',
      refreshToken: 'erneuerung-neu',
    );
  }

  @override
  Future<PushResponse> push(
    PushPayload payload, {
    required String accessToken,
  }) async {
    _checkToken();
    lastPush = payload;

    var applied = 0;
    final conflicts = <String>[];

    for (final dto in payload.repertoires) {
      final existing = repertoires[dto.uuid];
      final wins =
          existing == null ||
          dto.updatedAt.isAfter(existing.updatedAt) ||
          (dto.updatedAt == existing.updatedAt &&
              dto.revision >= existing.revision);

      if (wins) {
        repertoires[dto.uuid] = dto;
        applied++;
      } else {
        conflicts.add(dto.uuid);
      }
    }

    // Auch hier gilt die Regel des echten Servers: der jüngere Stand bleibt.
    for (final dto in payload.progress) {
      final key = '${dto.repertoireUuid}|${dto.pathHash}';
      final existing = progress[key];
      if (existing == null || dto.updatedAt.isAfter(existing.updatedAt)) {
        progress[key] = dto;
      }
    }

    if (payload.profile case final incoming?) {
      final existing = profile;
      if (existing == null || incoming.updatedAt.isAfter(existing.updatedAt)) {
        profile = incoming;
      }
    }

    return PushResponse(
      serverTime: serverTime,
      appliedRepertoires: applied,
      appliedProgress: payload.progress.length,
      conflictedRepertoires: conflicts,
    );
  }

  @override
  Future<PullResponse> pull({
    required String accessToken,
    DateTime? since,
  }) async {
    _checkToken();
    lastSince = since;

    return PullResponse(
      serverTime: serverTime,
      repertoires: repertoires.values.toList(),
      progress: progress.values.toList(),
      profile: profile,
    );
  }
}

void main() {
  late AppDatabase db;
  late _FakeApi api;
  late InMemorySyncSessionStore sessions;
  late SyncRepository repo;

  setUp(() async {
    db = AppDatabase.withExecutor(NativeDatabase.memory());
    api = _FakeApi();
    sessions = InMemorySyncSessionStore();
    repo = SyncRepository(db: db, api: api, sessions: sessions);

    await repo.signIn('lichess-token');
  });

  tearDown(() => db.close());

  Future<int> addRepertoire({
    String uuid = 'repertoire-0001',
    String name = 'Sizilianisch',
    DateTime? at,
  }) {
    return db.repertoireDao.insertRepertoire(
      uuid: uuid,
      name: name,
      side: Side.black,
      pgn: '[Event "?"]\n\n1. e4 c5 *',
      startFen: 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1',
      source: RepertoireSource.manual,
      nodeCount: 2,
      lineCount: 1,
      now: at ?? DateTime.utc(2026, 7, 29),
    );
  }

  group('Anmeldung', () {
    test('legt die Sitzung ab', () async {
      expect(await repo.isSignedIn, isTrue);
      expect((await sessions.read())!.accessToken, 'zugriff');
    });

    test('Abmelden räumt Sitzung und Marke weg', () async {
      await addRepertoire();
      await repo.sync();

      await repo.signOut();

      expect(await repo.isSignedIn, isFalse);
      expect(
        await db.keyValueDao.getDateTime(KeyValueDao.lastSyncAt),
        isNull,
      );
    });

    test('ohne Sitzung ist der Abgleich ein Fehler', () async {
      await repo.signOut();
      expect(repo.sync(), throwsA(isA<SyncException>()));
    });
  });

  group('Hinauf', () {
    test('schickt Repertoire, Lernstand und Spielerstand', () async {
      final id = await addRepertoire();
      await db.progressDao.ensureRows(
        repertoireId: id,
        pathHashes: ['a' * 40],
        now: DateTime.utc(2026, 7, 29),
      );

      final outcome = await repo.sync();

      expect(outcome.pushedRepertoires, 1);
      expect(outcome.pushedProgress, 1);
      expect(api.repertoires.keys, contains('repertoire-0001'));

      // Der Lernstand reist über die uuid, nicht über die lokale Zeilen-ID.
      final sent = api.lastPush!.progress.single;
      expect(sent.repertoireUuid, 'repertoire-0001');
      expect(sent.pathHash, 'a' * 40);
    });

    test('beim zweiten Mal geht nur Neues hinaus', () async {
      await addRepertoire();
      await repo.sync();

      final second = await repo.sync();

      expect(second.pushedRepertoires, 0);
      expect(api.lastPush!.repertoires, isEmpty);
      expect(api.lastSince, isNotNull);
    });

    test('eine Änderung danach geht wieder mit', () async {
      final id = await addRepertoire();
      await repo.sync();

      await db.repertoireDao.rename(id, 'Sizilianisch Najdorf');
      final outcome = await repo.sync();

      expect(outcome.pushedRepertoires, 1);
      expect(api.repertoires['repertoire-0001']!.name, 'Sizilianisch Najdorf');
    });

    test('ein gelöschtes Repertoire reist als Grabstein', () async {
      final id = await addRepertoire();
      await repo.sync();

      await db.repertoireDao.softDelete(id);
      await repo.sync();

      expect(api.repertoires['repertoire-0001']!.deletedAt, isNotNull);
    });
  });

  group('Herunter', () {
    test('legt ein fremdes Repertoire lokal an', () async {
      api.repertoires['vom-anderen-geraet'] = RepertoireDto(
        uuid: 'vom-anderen-geraet',
        name: 'Englisch',
        side: 'white',
        pgn: '[Event "?"]\n\n1. c4 *',
        startFen: 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1',
        source: 'manual',
        revision: 1,
        updatedAt: DateTime.utc(2026, 7, 30),
      );

      final outcome = await repo.sync();

      expect(outcome.pulledRepertoires, 1);
      final local = await db.repertoireDao.getByUuid('vom-anderen-geraet');
      expect(local!.name, 'Englisch');
      expect(local.side, Side.white);
    });

    test('ein älterer Serverstand überschreibt nichts', () async {
      final id = await addRepertoire(at: DateTime.utc(2026, 7, 31));
      final local = (await db.repertoireDao.getById(id))!;

      api.repertoires['repertoire-0001'] = RepertoireDto(
        uuid: local.uuid,
        name: 'alte Fassung',
        side: 'black',
        pgn: local.pgn,
        startFen: local.startFen,
        source: 'manual',
        updatedAt: DateTime.utc(2026, 7),
      );

      await repo.sync();

      expect((await db.repertoireDao.getById(id))!.name, 'Sizilianisch');
    });

    test('der Lernstand landet am richtigen Repertoire', () async {
      await addRepertoire();

      api.progress['repertoire-0001|${'b' * 40}'] = ProgressDto(
        repertoireUuid: 'repertoire-0001',
        pathHash: 'b' * 40,
        state: 'review',
        stability: 7.5,
        difficulty: 5,
        due: DateTime.utc(2026, 8, 5),
        reps: 4,
        lapses: 1,
        correctCount: 4,
        wrongCount: 1,
        updatedAt: DateTime.utc(2026, 7, 30),
      );

      final outcome = await repo.sync();
      expect(outcome.pulledProgress, 1);

      final id = (await db.repertoireDao.getByUuid('repertoire-0001'))!.id;
      final row = await db.progressDao.getOne(id, 'b' * 40);
      expect(row!.stability, 7.5);
      expect(row.state, FsrsState.review);
    });

    test('Lernstand ohne bekanntes Repertoire wird übergangen', () async {
      api.progress['fremd|${'c' * 40}'] = ProgressDto(
        repertoireUuid: 'gibt-es-nicht',
        pathHash: 'c' * 40,
        state: 'review',
        stability: 1,
        difficulty: 5,
        due: DateTime.utc(2026, 8, 5),
        reps: 1,
        lapses: 0,
        correctCount: 1,
        wrongCount: 0,
        updatedAt: DateTime.utc(2026, 7, 30),
      );

      final outcome = await repo.sync();
      expect(outcome.pulledProgress, 0);
    });

    test('übernimmt einen jüngeren Spielerstand', () async {
      api.profile = ProfileDto(
        totalXp: 5000,
        streakCurrent: 9,
        streakBest: 14,
        streakFreezes: 2,
        lastActiveDay: '2026-07-30',
        updatedAt: DateTime.utc(2027),
      );

      await repo.sync();

      final profile = await db.activityDao.profile();
      expect(profile.totalXp, 5000);
      expect(profile.streakBest, 14);
    });

    test('ein älterer Spielerstand bleibt liegen', () async {
      await db.activityDao.updateProfile(
        const UserProfilesCompanion(totalXp: Value(999)),
      );

      api.profile = ProfileDto(
        totalXp: 1,
        streakCurrent: 0,
        streakBest: 0,
        streakFreezes: 2,
        updatedAt: DateTime.utc(2020),
      );

      await repo.sync();

      expect((await db.activityDao.profile()).totalXp, 999);
    });
  });

  group('Abgelaufenes Token', () {
    test('wird einmal erneuert, danach läuft der Abgleich durch', () async {
      await addRepertoire();
      api.rejectFirstCall = true;

      final outcome = await repo.sync();

      expect(api.refreshCalls, 1);
      expect(outcome.pushedRepertoires, 1);
      expect((await sessions.read())!.accessToken, 'zugriff-neu');
    });
  });
}

File: backend/test/sync.test.ts
import { afterEach, beforeEach, describe, expect, it } from 'vitest';

import { incomingWins } from '../src/lib/sync.js';
import {
  bearer,
  createTestApp,
  progressPayload,
  repertoirePayload,
  signIn,
  type TestContext,
} from './helpers.js';

let context: TestContext;
let session: { accessToken: string; userId: string };

beforeEach(async () => {
  context = await createTestApp();
  session = await signIn(context);
});

afterEach(async () => {
  await context.close();
});

async function push(payload: unknown) {
  return context.app.inject({
    method: 'POST',
    url: '/v1/sync',
    headers: bearer(session.accessToken),
    payload: payload as Record<string, unknown>,
  });
}

async function pull(since?: string) {
  return context.app.inject({
    method: 'GET',
    url: since === undefined ? '/v1/sync' : `/v1/sync?since=${since}`,
    headers: bearer(session.accessToken),
  });
}

describe('Konfliktregel', () => {
  it('der jüngere Stand gewinnt', () => {
    const older = { updatedAt: new Date('2026-07-01') };
    const newer = { updatedAt: new Date('2026-07-02') };

    expect(incomingWins(newer, older)).toBe(true);
    expect(incomingWins(older, newer)).toBe(false);
  });

  it('bei gleicher Zeit entscheidet die Revision', () => {
    const at = new Date('2026-07-01');

    expect(
      incomingWins({ updatedAt: at, revision: 5 }, { updatedAt: at, revision: 4 }),
    ).toBe(true);
    expect(
      incomingWins({ updatedAt: at, revision: 3 }, { updatedAt: at, revision: 4 }),
    ).toBe(false);
  });

  it('eine unbekannte Zeile wird immer angenommen', () => {
    expect(incomingWins({ updatedAt: new Date() }, null)).toBe(true);
  });
});

describe('Hochladen', () => {
  it('nimmt Repertoire, Lernstand und Spielerstand an', async () => {
    const response = await push({
      repertoires: [repertoirePayload()],
      progress: [progressPayload()],
      profile: {
        totalXp: 1200,
        streakCurrent: 4,
        streakBest: 9,
        streakFreezes: 1,
        lastActiveDay: '2026-07-30',
        updatedAt: '2026-07-30T10:00:00.000Z',
      },
    });

    expect(response.statusCode).toBe(200);
    expect(response.json()).toMatchObject({
      applied: { repertoires: 1, progress: 1, profile: true },
    });

    expect(await context.prisma.repertoire.count()).toBe(1);
    expect(await context.prisma.nodeProgress.count()).toBe(1);
  });

  it('überschreibt mit einem jüngeren Stand', async () => {
    await push({ repertoires: [repertoirePayload()] });
    await push({
      repertoires: [
        repertoirePayload({
          name: 'Sizilianisch Najdorf',
          revision: 2,
          updatedAt: '2026-07-31T10:00:00.000Z',
        }),
      ],
    });

    const row = await context.prisma.repertoire.findFirstOrThrow();
    expect(row.name).toBe('Sizilianisch Najdorf');
    expect(await context.prisma.repertoire.count()).toBe(1);
  });

  it('meldet einen älteren Stand als Konflikt, statt ihn zu übernehmen', async () => {
    await push({
      repertoires: [
        repertoirePayload({
          name: 'neu',
          updatedAt: '2026-07-31T10:00:00.000Z',
        }),
      ],
    });

    const response = await push({
      repertoires: [
        repertoirePayload({
          name: 'alt',
          updatedAt: '2026-07-29T10:00:00.000Z',
        }),
      ],
    });

    expect(response.json()).toMatchObject({
      applied: { repertoires: 0 },
      conflicts: { repertoires: ['repertoire-0001'] },
    });

    const row = await context.prisma.repertoire.findFirstOrThrow();
    expect(row.name).toBe('neu');
  });

  it('behält ein gelöschtes Repertoire als Grabstein', async () => {
    await push({ repertoires: [repertoirePayload()] });
    await push({
      repertoires: [
        repertoirePayload({
          revision: 2,
          updatedAt: '2026-07-31T10:00:00.000Z',
          deletedAt: '2026-07-31T10:00:00.000Z',
        }),
      ],
    });

    const row = await context.prisma.repertoire.findFirstOrThrow();
    expect(row.deletedAt).not.toBeNull();
  });

  it('weist eine zu große Ladung ab', async () => {
    const response = await push({
      repertoires: Array.from({ length: 201 }, (_, index) =>
        repertoirePayload({ uuid: `repertoire-${index}` }),
      ),
    });

    expect(response.statusCode).toBe(400);
  });

  it('weist einen unbrauchbaren Lernstand ab', async () => {
    const response = await push({
      progress: [progressPayload({ pathHash: 'zu-kurz' })],
    });

    expect(response.statusCode).toBe(400);
  });
});

describe('Abholen', () => {
  it('liefert ohne Marke alles', async () => {
    await push({
      repertoires: [repertoirePayload()],
      progress: [progressPayload()],
    });

    const response = await pull();
    const body = response.json<{
      repertoires: unknown[];
      progress: unknown[];
      serverTime: string;
    }>();

    expect(body.repertoires).toHaveLength(1);
    expect(body.progress).toHaveLength(1);
    expect(Date.parse(body.serverTime)).not.toBeNaN();
  });

  it('liefert mit Marke nur, was seitdem dazukam', async () => {
    await push({ repertoires: [repertoirePayload()] });

    const mark = (await pull()).json<{ serverTime: string }>();

    const empty = (await pull(mark.serverTime)).json<{
      repertoires: unknown[];
    }>();
    expect(empty.repertoires).toHaveLength(0);

    await push({
      repertoires: [
        repertoirePayload({
          uuid: 'repertoire-0002',
          name: 'Englisch',
          side: 'white',
        }),
      ],
    });

    const after = (await pull(mark.serverTime)).json<{
      repertoires: { uuid: string }[];
    }>();
    expect(after.repertoires.map((r) => r.uuid)).toEqual(['repertoire-0002']);
  });

  it('gibt die Daten unverändert zurück', async () => {
    const sent = repertoirePayload();
    await push({ repertoires: [sent] });

    const body = (await pull()).json<{
      repertoires: Record<string, unknown>[];
    }>();

    expect(body.repertoires[0]).toMatchObject({
      uuid: sent.uuid,
      name: sent.name,
      side: sent.side,
      pgn: sent.pgn,
      updatedAt: sent.updatedAt,
    });
  });

  it('zeigt keine Daten eines anderen Kontos', async () => {
    await push({ repertoires: [repertoirePayload()] });

    context.identity = { id: 'jemand-anderes', username: 'Fremd' };
    const other = await signIn(context, 'anderes-token');

    const response = await context.app.inject({
      method: 'GET',
      url: '/v1/sync',
      headers: bearer(other.accessToken),
    });

    expect((response.json<{ repertoires: unknown[] }>()).repertoires).toEqual(
      [],
    );
  });

  it('verlangt eine Anmeldung', async () => {
    const response = await context.app.inject({
      method: 'GET',
      url: '/v1/sync',
    });
    expect(response.statusCode).toBe(401);
  });
});
